relay-server/src/actors/project.rs

```diff
@@ -606,24 +607,92 @@ impl Project {
         });
     }
 
+    fn ratelimit_and_merge_buckets(
+        &self,
+        project_state: Arc<ProjectState>,
+        mut buckets: Vec<Bucket>,
+        aggregator: Addr<Aggregator>,
+        #[allow(unused_variables)] envelope_processor: Addr<EnvelopeProcessor>,
+        outcome_aggregator: Addr<TrackOutcome>,
+    ) {
+        let Some(scoping) = self.scoping() else {
+            // This shouldn't be possible since we check if project id is present before setting the new project state.
+            // TODO(tor): Find a way to properly represent this invariant in the type system.
+            relay_log::trace!(
+                "there is no scoping due to missing project id: dropping {} buckets",
+                buckets.len()
+            );
+            return;
+        };
+
+        // Only send if the project state is valid, otherwise drop the buckets.
+        if project_state.check_disabled(self.config.as_ref()).is_err() {
+            relay_log::trace!("project state invalid: dropping {} buckets", buckets.len());
+            return;
+        }
+
+        // Re-run feature flag checks since the project might not have been loaded when the buckets
+        // were initially ingested, or feature flags have changed in the meanwhile.
+        self.filter_metrics(&mut buckets);
+
+        // Check rate limits if necessary:
+        let quotas = project_state.config.quotas.clone();
+
+        let usage = match project_state.config.transaction_metrics {
+            Some(ErrorBoundary::Ok(ref c)) => c.usage_metric(),
+            _ => false,
+        };
+
+        let buckets = match MetricsLimiter::create(buckets, quotas, scoping, usage) {
+            Ok(mut bucket_limiter) => {
+                let cached_rate_limits = self.rate_limits().clone();
+                #[allow(unused_variables)]
+                let was_rate_limited =
+                    bucket_limiter.enforce_limits(Ok(&cached_rate_limits), outcome_aggregator);
+
+                #[cfg(feature = "processing")]
+                if !was_rate_limited && self.config.processing_enabled() {
+                    // If there were no cached rate limits active, let the processor check redis:
+                    envelope_processor.send(RateLimitFlushBuckets { bucket_limiter });
```

Comment:
Let's rename to `RateLimitBuckets` -- "flush" seems to imply they were already flushed by the aggregator.